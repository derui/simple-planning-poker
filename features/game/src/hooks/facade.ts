import { createFacade } from "@spp/shared-hook-facade";
import { UseCreateGame, UseListGame, UsePrepareGame } from "../atoms/game.js";

export type Hooks = {
  useCreateGame: UseCreateGame;
  usePrepareGame: UsePrepareGame;
  useListGame: UseListGame;
};

const { hooks, ImplementationProvider } = createFacade<Hooks>();

export { hooks, ImplementationProvider };
