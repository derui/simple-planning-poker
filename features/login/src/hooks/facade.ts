import { createFacade } from "@spp/shared-hook-facade";
import { UseAuth, UseLogin } from "../atoms/atom.js";

export type Hooks = {
  useAuth: UseAuth;
  useLogin: UseLogin;
};

const facade: ReturnType<typeof createFacade<Hooks>> = createFacade<Hooks>();

export const hooks: Readonly<Hooks> = facade.hooks;
export const ImplementationProvider: typeof facade.ImplementationProvider = facade.ImplementationProvider;
