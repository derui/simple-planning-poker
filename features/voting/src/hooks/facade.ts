import { createFacade } from "@spp/shared-hook-facade";
import { UseJoin, UsePollingPlace, UseRevealed, UseVoting, UseVotingStatus } from "../atoms/voting.js";

export type Hooks = {
  useJoin: UseJoin;
  useVoting: UseVoting;
  useRevealed: UseRevealed;
  useVotingStatus: UseVotingStatus;
  usePollingPlace: UsePollingPlace;
};

const facade: ReturnType<typeof createFacade<Hooks>> = createFacade<Hooks>();

export const hooks: Readonly<Hooks> = facade.hooks;
export const ImplementationProvider: typeof facade.ImplementationProvider = facade.ImplementationProvider;
