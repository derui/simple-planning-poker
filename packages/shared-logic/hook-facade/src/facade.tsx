import { createContext, useContext } from "react";

export type ImplementationProvider<T> = React.ComponentType<React.PropsWithChildren<{ implementation: T }>>;

/**
 * Create facade with type `Hooks`
 */
export const createFacade = function createFacade<Hooks extends Record<string, () => unknown>>(): {
  hooks: Readonly<Hooks>;
  ImplementationProvider: ImplementationProvider<Hooks>;
} {
  const providerNotFound: unique symbol = Symbol();

  const Context = createContext<Hooks | typeof providerNotFound>(providerNotFound);

  const implementationHandler: ProxyHandler<Hooks> = {
    get(_target, prop, _receiver) {
      // eslint-disable-next-line
      const obj = useContext(Context);

      if (obj === providerNotFound) {
        throw new Error("Provider not found");
      }

      if (typeof prop == "symbol") {
        throw new Error("Can not use symbol as hook key");
      }

      return obj[prop];
    },
    has() {
      return false;
    },
    ownKeys() {
      return [];
    },
    getOwnPropertyDescriptor() {
      return undefined;
    },
    getPrototypeOf() {
      return null;
    },
    preventExtensions() {
      return true;
    },
    isExtensible() {
      return false;
    },
    set() {
      return false;
    },
    deleteProperty() {
      return false;
    },
  };

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const proxy = new Proxy<Hooks>({} as Hooks, implementationHandler);

  const ImplementationProvider: ImplementationProvider<Hooks> = ({ implementation, children }) => {
    const consumed = useContext(Context);

    if (consumed !== providerNotFound) {
      throw new Error("Provider already defined");
    }

    return <Context.Provider value={implementation}>{children}</Context.Provider>;
  };

  return { hooks: proxy, ImplementationProvider };
};
