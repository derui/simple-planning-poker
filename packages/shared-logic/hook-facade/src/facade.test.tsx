import { expect, test } from "vitest";
import { createFacade } from "./facade.js";
import { renderHook } from "@testing-library/react";

test("pass implementation", () => {
  // Arrange
  type T = {
    useAString(): string;
  };

  const { hooks, ImplementationProvider } = createFacade<T>();
  const mock: T = {
    useAString() {
      return "foo";
    },
  };

  // Act
  const { result } = renderHook(() => hooks.useAString(), {
    wrapper: ({ children }) => <ImplementationProvider implementation={mock}>{children}</ImplementationProvider>,
  });

  // Assert
  expect(result.current).toEqual("foo");
});

test("should throw error when do not provide implementation", () => {
  // Arrange
  type T = {
    useAString(): string;
  };

  const { hooks, ImplementationProvider } = createFacade<T>();

  // Act

  // Assert
  expect(() => renderHook(() => hooks.useAString())).toThrowErrorMatchingSnapshot();
});
