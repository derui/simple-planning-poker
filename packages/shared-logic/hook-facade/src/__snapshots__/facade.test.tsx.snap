// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should throw error when do not provide implementation 1`] = `[Error: Provider not found]`;
